### PR TITLE
Introduce a mailer interface and the respective factory

### PR DESCRIPTION
**Pull request by laoneo**
### Summary of Changes
Introduces a new `MailerInterface` which does contain the most common functions which are needed to send a mail. A factory which is served through the container and the corresponding aware interface and trait are also shipped with this pr.

It also fixes an inconsistency, getting a mailer with the global configuration settings. `Mail::getInstance()` delivers the same object, but with a different configuration, either if `Factory::getMailer()` is called before or not. Here is an example:

```
Factory::getConfig()->set('mailfrom', 'test@example.com');
echo Mail::getInstance()->From; // Prints an empty string
echo Factory::getMailer()->From; // Prints test@example.com
echo Mail::getInstance()->From; // Prints now test@example.com
```

With this pr all calls do return as From value the EMail address from the global configuration.

### Testing Instructions
Send a test mail while editing the global Joomla configuration.

### Actual result BEFORE applying this Pull Request
Test mail is sent.

### Expected result AFTER applying this Pull Request
Test mail is sent.

### Link to documentations
Please select:
- [ ] Documentation link for docs.joomla.org: <link>
- [x] No documentation changes for docs.joomla.org needed

- [x] Pull Request link for manual.joomla.org: https://github.com/joomla/Manual/pull/116
- [ ] No documentation changes for manual.joomla.org needed

### CODE REVIEW

**Comment by wilsonge** [comment]
I don't think this is the correct implementation for this.

Note that factory doesn't actually set the mailer into it's list of statics - all it's doing is populating the data from JApplication. Funnily enough as things stand if all applications only called `Mail::getInstance` instead of `Factory::getMailer` then they wouldn't have this data populated despite the doc blocks. Now all instances of the mailer will have this information injected if they are using the factory which is a big behaviour change.

This is a much simpler case than the mvc classes. I think you just need an interface on the mail class and on top of that a new method in the application that bootstraps a mail instance that can be injected. This will also make the application dependent code much easier to unit test and from the models perspective it makes no diference anyhow whether that bootstrap of the from sender etc has actually taken place or not when the mail object is being injected.

**Comment by laoneo** [comment]
More or less I was there at the beginning. Then I though that mostly you want to have the settings in the mailer instance from the global config.

> Note that factory doesn't actually set the mailer into it's list of statics - all it's doing is populating the data from JApplication

This is not correct, it calls `Mail::getInstance()` which is cached, so it basically sets that config on the cached instance. A subsequent call on `Mail::getInstance()` delivers a mailer with the global settings. So actually this is a pretty inconsistent behaviour as you can get two different mailer, depending if `Factory::getMailer()` is called already or not. 

> a new method in the application

Please no more functions which do make the application instance an even bigger god object.

**Comment by laoneo** [comment]
> OK but what is the value of having that central object stored in the container? Like the container is there to do dependency management not configuration of objects

Only the factory is in the container, not the mailer instance.

**Comment by ChrisHoefliger** [comment]
I have tested this unsuccessfully on localhost /w mailer set to SMPT
mailerror

**Comment by laoneo** [comment]
Does it work without the patch? Just want to make sure that the settings itself do work.

**Comment by ChrisHoefliger** [comment]
> Does it work without the patch? Just want to make sure that the settings itself do work.

Yeah, I tried without the patch first, that worked, then with patch applied and finally with patch removed. Only with applied patch no luck. This is the latest nighly build on PHP 8.2.1.

**Comment by laoneo** [comment]
Did you try with the send test mail button?

**Comment by ChrisHoefliger** [comment]
> Does it work without the patch? Just want to make sure that the settings itself do work.

It says 'Send a test mail while editing the global Joomla configuration.'
I assume that means press the send Test Mail button - that's what I did. Without the patch it says 'Test mail sent'; applying the patch, there's an error message and no mail gets sent.

File: libraries/src/Mail/MailerInterface.php [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+ * @copyright  (C) 2023 Open Source Matters, Inc. <https://www.joomla.org>
+ * @license    GNU General Public License version 2 or later; see LICENSE.txt
+ */
+
+namespace Joomla\CMS\Mail;
+
+use RuntimeException;
+
+// phpcs:disable PSR1.Files.SideEffects
+\defined('JPATH_PLATFORM') or die;
```
Comment (heelc29) [inline libraries/src/Mail/MailerInterface.php]:
```suggestion
\defined('_JEXEC') or die;
```

File: libraries/src/Mail/MailerFactory.php [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+     * The default configuration.
+     *
+     * @var  Registry
+     */
+    private $defaultConfiguration;
+
+    /**
+     * The MailerFactory constructor.
+     *
+     * @param   Registry  $db  The database
```
Comment (heelc29) [inline libraries/src/Mail/MailerFactory.php]:
```suggestion
     * @param   Registry  $defaultConfiguration  The default configuration
```
or something else...

File: libraries/src/Mail/MailerFactory.php [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+ * Default factory for creating mailer objects.
+ *
+ * @since  __DEPLOY_VERSION__
+ */
+class MailerFactory implements MailerFactoryInterface
+{
+    /**
+     * The default configuration.
+     *
+     * @var  Registry
```
Comment (heelc29) [inline libraries/src/Mail/MailerFactory.php]:
```suggestion
     * @var     Registry
     * @since   __DEPLOY_VERSION__
```

File: libraries/src/Mail/MailerFactory.php [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+     *
+     * @since   __DEPLOY_VERSION__
+     */
+    public function createMailer(Registry $configuration = null): MailerInterface
+    {
+        if ($configuration === null) {
+            $configuration = $this->defaultConfiguration;
+        }
+
+        $mailer = new Mail((bool)$configuration->get('throw_exceptions', true));
```
Comment (heelc29) [inline libraries/src/Mail/MailerFactory.php]:
```suggestion
        $mailer = new Mail((bool) $configuration->get('throw_exceptions', true));
```

File: libraries/src/Mail/MailerFactoryAwareTrait.php [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+
+        throw new \UnexpectedValueException('MailerFactory not set in ' . __CLASS__);
+    }
+
+    /**
+     * Set the mailer factory to use.
+     *
+     * @param   ?MailerFactoryInterface  $mailerFactory  The mailer factory to use.
+     *
+     * @return  $this
```
Comment (heelc29) [inline libraries/src/Mail/MailerFactoryAwareTrait.php]:
```suggestion
     * @return  void
```

File: libraries/src/Mail/MailerFactoryAwareTrait.php [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+     *
+     * @since   __DEPLOY_VERSION__
+     */
+    public function setMailerFactory(MailerFactoryInterface $mailerFactory = null): void
```
Comment (heelc29) [inline libraries/src/Mail/MailerFactoryAwareTrait.php]: `?MailerFactoryInterface` or `MailerFactoryInterface $xxx = null`

**Comment by laoneo** [comment]
@ChrisHoefliger can you delete the file /administrator/cache/autoload_psr4.php? Does it then still not work?

File: libraries/src/Mail/MailerFactoryAwareTrait.php (hunk 6 above)
Comment (laoneo) [reply libraries/src/Mail/MailerFactoryAwareTrait.php]: ?MailerFactoryInterface :smirk:

**Comment by ChrisHoefliger** [comment]
> @ChrisHoefliger can you delete the file /administrator/cache/autoload_psr4.php? Does it then still not work?

After deleting autoload_psr4.php and then re-applying this patch, it still does not work. Without the patch, it works.

**Comment by ChrisHoefliger** [comment]
I have now re-tested this successfully on a test site, using SMTP mail on PHP 8.2

**Comment by ChrisHoefliger** [comment]
I have tested this item :white_check_mark: successfully on a4df9b9e7c69ec52d12033f230028063b70d5a1a<br /><br />Tested successfully with SMTP<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/40560">issues.joomla.org/tracker/joomla-cms/40560</a>.</sub>

**Comment by toivo** [comment]
I have tested this item :white_check_mark: successfully on a4df9b9e7c69ec52d12033f230028063b70d5a1a<br /><br />Tested successfully in Joomla 4.4.0-dev of 28 May in Wampserver 3.3.1 64-bit on Windows 11 using SMTP and PHP 8.1.6.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/40560">issues.joomla.org/tracker/joomla-cms/40560</a>.</sub>

**Comment by alikon** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/40560">issues.joomla.org/tracker/joomla-cms/40560</a>.</sub>

File: libraries/src/Mail/MailerFactoryAwareInterface.php [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+    /**
+     * Set the mailer factory to use.
+     *
+     * @param   ?MailerFactoryInterface  $factory  The mailer factory to use
+     *
+     * @return  void
+     *
+     * @since   __DEPLOY_VERSION__
+     */
+    public function setMailerFactory(?MailerFactoryInterface $factory): void;
```
Comment (heelc29) [inline libraries/src/Mail/MailerFactoryAwareInterface.php]:
```suggestion
    /**
     * Set the mailer factory to use.
     *
     * @param   ?MailerFactoryInterface  $mailerFactory  The mailer factory to use.
     *
     * @return  void
     *
     * @since   __DEPLOY_VERSION__
     */
    public function setMailerFactory(?MailerFactoryInterface $mailerFactory): void;
```
same variable name as used in trait

File: libraries/src/Mail/MailerFactory.php [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+     * Method to get an instance of a mailer. If the passed settings are null,
+     * then the mailer does use the internal configuration.
+     *
+     * @param   Registry  $settings  The configuration
+     *
+     * @return  MailerInterface
+     *
+     * @since   __DEPLOY_VERSION__
+     */
+    public function createMailer(Registry $settings = null): MailerInterface
```
Comment (heelc29) [inline libraries/src/Mail/MailerFactory.php]:
```suggestion
     * @param   ?Registry  $settings  The configuration
     *
     * @return  MailerInterface
     *
     * @since   __DEPLOY_VERSION__
     */
    public function createMailer(?Registry $settings): MailerInterface
```

File: libraries/src/Mail/MailerFactoryInterface.php [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+     * Method to get an instance of a mailer. If the passed settings are null,
+     * then the mailer does use the internal configuration.
+     *
+     * @param   Registry  $settings  The configuration
+     *
+     * @return  MailerInterface
+     *
+     * @since   __DEPLOY_VERSION__
+     */
+    public function createMailer(Registry $settings = null): MailerInterface;
```
Comment (heelc29) [inline libraries/src/Mail/MailerFactoryInterface.php]:
```suggestion
     * @param   ?Registry  $settings  The configuration
     *
     * @return  MailerInterface
     *
     * @since   __DEPLOY_VERSION__
     */
    public function createMailer(?Registry $settings): MailerInterface;
```

**Comment by heelc29** [comment]
ParamSignatureMismatch of `MailerInterface` to `Joomla\CMS\Mail\Mail`
```
libraries/src/Mail/Mail.php:134 PhanParamSignatureMismatch Declaration of function Send() : bool should be compatible with function send() : void defined in libraries/src/Mail/MailerInterface.php:34
libraries/src/Mail/Mail.php:197 PhanParamSignatureMismatch Declaration of function setSender(mixed $from, mixed|string $name = '') : \Joomla\CMS\Mail\Mail|bool should be compatible with function setSender(string $fromEmail, string $name = '') : void defined in libraries/src/Mail/MailerInterface.php:48
libraries/src/Mail/Mail.php:233 PhanParamSignatureMismatch Declaration of function setSubject(string $subject) : \Joomla\CMS\Mail\Mail should be compatible with function setSubject(string $subject) : void defined in libraries/src/Mail/MailerInterface.php:59
libraries/src/Mail/Mail.php:249 PhanParamSignatureMismatch Declaration of function setBody(string $content) : \Joomla\CMS\Mail\Mail should be compatible with function setBody(string $content) : void defined in libraries/src/Mail/MailerInterface.php:70
libraries/src/Mail/Mail.php:332 PhanParamSignatureMismatch Declaration of function addRecipient(mixed $recipient, mixed|string $name = '') : \Joomla\CMS\Mail\Mail|bool should be compatible with function addRecipient(string $recipientEmail, string $name = '') : void defined in libraries/src/Mail/MailerInterface.php:84
libraries/src/Mail/Mail.php:349 PhanParamSignatureMismatch Declaration of function addCc(mixed $cc, mixed|string $name = '') : \Joomla\CMS\Mail\Mail|bool should be compatible with function addCc(string $ccEmail, string $name = '') : void defined in libraries/src/Mail/MailerInterface.php:98
libraries/src/Mail/Mail.php:371 PhanParamSignatureMismatch Declaration of function addBcc(mixed $bcc, mixed|string $name = '') : \Joomla\CMS\Mail\Mail|bool should be compatible with function addBcc(string $bccEmail, string $name = '') : void defined in libraries/src/Mail/MailerInterface.php:112
libraries/src/Mail/Mail.php:397 PhanParamSignatureMismatch Declaration of function addAttachment(mixed $path, mixed|string $name = '', mixed|string $encoding = 'base64', mixed|string $type = 'application/octet-stream', string $disposition = 'attachment') : \Joomla\CMS\Mail\Mail|bool should be compatible with function addAttachment(string $data, string $name = '', string $encoding = 'base64', string $type = 'application/octet-stream') : void defined in libraries/src/Mail/MailerInterface.php:126
libraries/src/Mail/Mail.php:481 PhanParamSignatureMismatch Declaration of function addReplyTo(mixed $replyto, mixed|string $name = '') : \Joomla\CMS\Mail\Mail|bool should be compatible with function addReplyTo(string $replyToEmail, string $name = '') : void defined in libraries/src/Mail/MailerInterface.php:140
```

**Comment by laoneo** [comment]
Then phan messages can be ignored for now as the interface should be decoupled from the actual Mail implementation. Till the interface gets wider acceptance and we can change the methods in mailer we have to leave with the mismatch. PHP itself is accepting that and this is what counts for me at the moment. The target is to have an independent mailer interface without any reference to an implementation so we can exchange it on some point when there is a need for it.

File: libraries/src/Mail/Mail.php [hunk 10]
```diff
@@ -183,7 +194,7 @@ public function Send()
      * @throws  \UnexpectedValueException  if the sender is not a valid address
      * @throws  phpmailerException          if setting the sender failed and exception throwing is enabled
      */
-    public function setSender($from)
+    public function setSender($from, $name = '')
```
Comment (HLeithner) [inline libraries/src/Mail/Mail.php:197]: seems unrelated

File: libraries/src/Mail/MailerFactory.php [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+ */
+class MailerFactory implements MailerFactoryInterface
+{
+    /**
+     * The default configuration.
+     *
+     * @var     Registry
+     * @since   __DEPLOY_VERSION__
+     */
+    private $defaultConfiguration;
```
Comment (HLeithner) [inline libraries/src/Mail/MailerFactory.php:33]: Adding a type hint could make sense.

File: libraries/src/Mail/MailerInterface.php [hunk 12]
```diff
@@ -0,0 +1,141 @@
...
+    /**
+     * Set the email body.
+     *
+     * @param   string  $content  Body of the email
+     *
+     * @return  void
+     *
+     * @since   __DEPLOY_VERSION__
+     */
+    public function setBody(string $content);
```
Comment (HLeithner) [inline libraries/src/Mail/MailerInterface.php:70]: Is this the email or plain text body? Would it make sense to have 2 functions? iirc you have to set the variable `AltBody` for the plaintext body.

File: libraries/src/Mail/Mail.php (hunk 10 above)
Comment (laoneo) [reply libraries/src/Mail/Mail.php:197]: Not really, why?

File: libraries/src/Mail/MailerFactory.php (hunk 11 above)
Comment (laoneo) [reply libraries/src/Mail/MailerFactory.php:33]: Then it's not 7.2 compatible.

File: libraries/src/Mail/MailerInterface.php (hunk 12 above)
Comment (laoneo) [reply libraries/src/Mail/MailerInterface.php:70]: You can either set HTML or plain text.

Comment (HLeithner) [reply libraries/src/Mail/MailerInterface.php:70]: how should I set a html content and a plain text content with the same function without a parameter?
